File: src/main/java/me/shedaniel/linkie/spring/FabricRoute.java
package me.shedaniel.linkie.spring;

import com.vaadin.flow.component.checkbox.Checkbox;
import com.vaadin.flow.component.html.H2;
import com.vaadin.flow.component.html.Label;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.select.Select;
import com.vaadin.flow.component.textfield.TextArea;
import com.vaadin.flow.router.Route;
import me.shedaniel.cursemetaapi.CurseMetaAPI;
import me.shedaniel.linkie.Pair;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Route("fabric")
public class FabricRoute extends VerticalLayout {
    private static final String BUILD_GRADLE_TEXT = "dependencies {\n    minecraft \"com.mojang:minecraft:@MC@\"\n    mappings \"@YARN@\"\n    modCompile \"@LOADER@\"\n    \n    // Optional Fabric API\n    modCompile \"@API@\"\n}";
    private static final String GRADLE_PROPERTIES_TEXT = "minecraft_version=@MC@\nyarn_mappings=@YARN@\nloader_version=@LOADER@\n\n# Optional Fabric API\nfabric_version=@API@";
    private Checkbox showSnapshots = new Checkbox("Show Snapshots");
    private TextArea buildGradleArea = new TextArea("build.gradle");
    private TextArea gradlePropertiesArea = new TextArea("gradle.properties (Example Mod)");
    private Select<LoadMeta.MinecraftVersion> versionSelect = new Select<>();
    private Label warning = new Label();
    
    public FabricRoute() {
        add(new H2("Fabric Version"));
        versionSelect.setItems(Collections.emptyList());
        versionSelect.setLabel("Minecraft Version: ");
        versionSelect.setItemLabelGenerator(item -> item.version);
        versionSelect.setRequiredIndicatorVisible(true);
        showSnapshots.addValueChangeListener(event -> updateComboBox(event.getValue()));
        versionSelect.addValueChangeListener(event -> updateInfo(versionSelect.getValue()));
        versionSelect.setEmptySelectionAllowed(false);
        versionSelect.setTextRenderer(item -> item.version);
        updateComboBox(false);
        {
            HorizontalLayout horizontalLayout = new HorizontalLayout(versionSelect, showSnapshots);
            horizontalLayout.setDefaultVerticalComponentAlignment(Alignment.END);
            add(horizontalLayout);
        }
        buildGradleArea.setWidthFull();
        buildGradleArea.setReadOnly(true);
        gradlePropertiesArea.setWidthFull();
        gradlePropertiesArea.setReadOnly(true);
        warning.getStyle().set("color", "red");
        add(warning);
        add(buildGradleArea);
        add(gradlePropertiesArea);
        updateInfo(versionSelect.getValue());
    }
    
    public void updateComboBox(boolean showSnapshots) {
        if (!showSnapshots) {
            versionSelect.setItems(LoadMeta.minecraftVersions.stream().filter(LoadMeta.MinecraftVersion::isStable).collect(Collectors.toList()));
        } else {
            versionSelect.setItems(Collections.unmodifiableList(LoadMeta.minecraftVersions));
        }
        versionSelect.setValue(LoadMeta.minecraftVersions.stream().filter(LoadMeta.MinecraftVersion::isStable).findFirst().get());
    }
    
    public void updateInfo(LoadMeta.MinecraftVersion version) {
        List<LoadMeta.MinecraftVersion> minecraftVersions = new ArrayList<>(LoadMeta.minecraftVersions);
        if (version == null)
            version = minecraftVersions.stream().filter(LoadMeta.MinecraftVersion::isStable).findFirst().get();
        String possibleFabricApiVersion = "@API@";
        String possibleCleanFabricApiVersion = "@API@";
        boolean firstOne = true;
        Label123:
        for(int i = minecraftVersions.indexOf(version); i < minecraftVersions.size(); i++) {
            String s = minecraftVersions.get(i).version;
            for(Map.Entry<String, Pair<CurseMetaAPI.AddonFile, Boolean>> entry : LoadMeta.fabricApi.entrySet()) {
                if (s.equalsIgnoreCase(entry.getKey())) {
                    if (entry.getValue().getKey().fileName.startsWith("fabric-api-")) {
                        possibleFabricApiVersion = "net.fabricmc.fabric-api:fabric-api:" + entry.getValue().getKey().fileName.replaceFirst("fabric-api-", "").replace(".jar", "");
                    } else {
                        possibleFabricApiVersion = "net.fabricmc:fabric:" + entry.getValue().getKey().fileName.replaceFirst("fabric-", "").replace(".jar", "");
                    }
                    possibleCleanFabricApiVersion = entry.getValue().getKey().fileName.replaceFirst("fabric-api-", "").replaceFirst("fabric-", "").replace(".jar", "");
                    firstOne = i == minecraftVersions.indexOf(version);
                    break Label123;
                }
            }
        }
        buildGradleArea.setValue(BUILD_GRADLE_TEXT
                .replaceAll("@MC@", version.version)
                .replaceAll("@YARN@", version.yarnMaven == null ? "@YARN@" : version.yarnMaven)
                .replaceAll("@LOADER@", LoadMeta.loaderVersion == null ? "@LOADER@" : LoadMeta.loaderVersion)
                .replaceAll("@API@", possibleFabricApiVersion)
        );
        gradlePropertiesArea.setValue(GRADLE_PROPERTIES_TEXT
                .replaceAll("@MC@", version.version)
                .replaceAll("@YARN@", version.yarnMaven == null ? "@YARN@" : version.yarnMaven.substring(version.yarnMaven.lastIndexOf(':') + 1))
                .replaceAll("@LOADER@", LoadMeta.loaderVersion == null ? "@LOADER@" : LoadMeta.loaderVersion.substring(LoadMeta.loaderVersion.lastIndexOf(':') + 1))
                .replaceAll("@API@", possibleCleanFabricApiVersion)
        );
        if (possibleFabricApiVersion.equals("@API@"))
            warning.setText("ERROR: This version DOES NOT have a fabric api version!");
        else
            warning.setText(firstOne ? "" : "WARNING: This version DOES NOT have a fabric api version! The closest fabric api version has been chosen. Please browse the version here: https://www.curseforge.com/minecraft/mc-mods/fabric-api/files/all");
    }
}

File: src/main/kotlin/me/shedaniel/linkie/spring/FabricRoute.kt
@file:Suppress("unused")

package me.shedaniel.linkie.spring

import com.vaadin.flow.component.checkbox.Checkbox
import com.vaadin.flow.component.html.H2
import com.vaadin.flow.component.html.Label
import com.vaadin.flow.component.orderedlayout.FlexComponent
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.component.select.Select
import com.vaadin.flow.component.textfield.TextArea
import com.vaadin.flow.router.Route

@Route("fabric")
class FabricRoute : VerticalLayout() {
    private val buildGradleText = "dependencies {\n    minecraft \"com.mojang:minecraft:@MC@\"\n    mappings \"@YARN@\"\n    modCompile \"@LOADER@\"\n    \n    // Optional Fabric API\n    modCompile \"@API@\"\n}"
    private val gradlePropertiesText = "minecraft_version=@MC@\nyarn_mappings=@YARN@\nloader_version=@LOADER@\n\n# Optional Fabric API\nfabric_version=@API@"
    private val showSnapshots = Checkbox("Show Snapshots")
    private val buildGradleArea = TextArea("build.gradle")
    private val gradlePropertiesArea = TextArea("gradle.properties (Example Mod)")
    private val versionSelect = Select<MinecraftVersion>()
    private val warning = Label()

    init {
        add(H2("Fabric Version"))
        versionSelect.setItems(emptyList<MinecraftVersion>())
        versionSelect.label = "Minecraft Version: "
        versionSelect.setItemLabelGenerator { it.version }
        versionSelect.isRequiredIndicatorVisible = true
        showSnapshots.addValueChangeListener { updateComboBox(it.value) }
        versionSelect.addValueChangeListener { updateInfo(versionSelect.value) }
        versionSelect.isEmptySelectionAllowed = false
        versionSelect.setTextRenderer { it.version }
        updateComboBox(false)

        val horizontalLayout = HorizontalLayout(versionSelect, showSnapshots)
        horizontalLayout.defaultVerticalComponentAlignment = FlexComponent.Alignment.END
        add(horizontalLayout)

        buildGradleArea.setWidthFull()
        buildGradleArea.isReadOnly = true
        gradlePropertiesArea.setWidthFull()
        gradlePropertiesArea.isReadOnly = true
        warning.style.set("color", "red")
        add(warning)
        add(buildGradleArea)
        add(gradlePropertiesArea)
        updateInfo(versionSelect.value)
    }

    private fun updateComboBox(showSnapshots: Boolean) {
        if (!showSnapshots) {
            versionSelect.setItems(minecraftVersions.filter(MinecraftVersion::stable))
        } else {
            versionSelect.setItems(minecraftVersions.toMutableList())
        }
        versionSelect.value = minecraftVersions.firstOrNull(MinecraftVersion::stable)
    }

    private fun updateInfo(versionNullable: MinecraftVersion?) {
        val minecraftVersions = minecraftVersions.toMutableList()
        val version = versionNullable ?: minecraftVersions.firstOrNull(MinecraftVersion::stable) ?: return
        var possibleFabricApiVersion = "@API@"
        var possibleCleanFabricApiVersion = "@API@"
        var firstOne = true
        Label123@ for (i in minecraftVersions.indexOf(version) until minecraftVersions.size) {
            val s = minecraftVersions[i].version
            for ((key, value) in fabricApi) {
                if (s.equals(key, ignoreCase = true)) {
                    possibleFabricApiVersion = if (value.first.fileName.startsWith("fabric-api-")) {
                        "net.fabricmc.fabric-api:fabric-api:" + value.first.fileName.replaceFirst("fabric-api-", "").replace(".jar", "")
                    } else {
                        "net.fabricmc:fabric:" + value.first.fileName.replaceFirst("fabric-", "").replace(".jar", "")
                    }
                    possibleCleanFabricApiVersion = value.first.fileName.replaceFirst("fabric-api-", "").replaceFirst("fabric-", "").replace(".jar", "")
                    firstOne = i == minecraftVersions.indexOf(version)
                    break@Label123
                }
            }
        }
        buildGradleArea.value = buildGradleText
                .replace("@MC@", version.version)
                .replace("@YARN@", if (version.yarnMaven == null) "@YARN@" else version.yarnMaven!!)
                .replace("@LOADER@", if (loaderVersion == null) "@LOADER@" else loaderVersion!!)
                .replace("@API@", possibleFabricApiVersion)
        gradlePropertiesArea.value = gradlePropertiesText
                .replace("@MC@", version.version)
                .replace("@YARN@", if (version.yarnMaven == null) "@YARN@" else version.yarnMaven!!.substring(version.yarnMaven!!.lastIndexOf(':') + 1))
                .replace("@LOADER@", if (loaderVersion == null) "@LOADER@" else loaderVersion!!.substring(loaderVersion!!.lastIndexOf(':') + 1))
                .replace("@API@", possibleCleanFabricApiVersion)
        if (possibleFabricApiVersion == "@API@")
            warning.text = "ERROR: This version DOES NOT have a fabric api version!"
        else
            warning.text = if (firstOne) "" else "WARNING: This version DOES NOT have a fabric api version! The closest fabric api version has been chosen. Please browse the version here: https://www.curseforge.com/minecraft/mc-mods/fabric-api/files/all"
    }
}